crypto/x509/v3_ncons.c

```diff
@@ -556,3 +561,118 @@ static int nc_uri(const ASN1_IA5STRING *uri, const ASN1_IA5STRING *base) {
 
   return X509_V_OK;
 }
+
+#define IPV4_ADDR_LEN 4
+#define IPV4_CIDR_LEN (IPV4_ADDR_LEN * 2)
+#define IPV6_ADDR_LEN 16
+#define IPV6_CIDR_LEN (IPV6_ADDR_LEN * 2)
+
+static int validate_ipv4_cidr_mask(const uint8_t *mask_ptr, size_t mask_len) {
+  assert(mask_len == IPV4_ADDR_LEN);
+
+  uint32_t mask = ((uint32_t)mask_ptr[0] << 24) |
+                  ((uint32_t)mask_ptr[1] << 16) | ((uint32_t)mask_ptr[2] << 8) |
+                  ((uint32_t)mask_ptr[3]);
+
+  if (mask != 0 && ((~mask + 1) & ~mask)) {
+    return 0;
+  }
+
+  return 1;
+}
+
+static int validate_ipv6_cidr_mask(const uint8_t *mask_ptr, size_t mask_len) {
+  assert(mask_len == IPV6_ADDR_LEN);
+
+  uint64_t mask_high =
+      (((uint64_t)mask_ptr[0]) << 56) | ((uint64_t)mask_ptr[1] << 48) |
+      ((uint64_t)mask_ptr[2] << 40) | ((uint64_t)mask_ptr[3] << 32) |
+      ((uint64_t)mask_ptr[4] << 24) | ((uint64_t)mask_ptr[5] << 16) |
+      ((uint64_t)mask_ptr[6] << 8) | ((uint64_t)mask_ptr[7]);
+
+  uint64_t mask_low =
+      (((uint64_t)mask_ptr[8]) << 56) | ((uint64_t)mask_ptr[9] << 48) |
+      ((uint64_t)mask_ptr[10] << 40) | ((uint64_t)mask_ptr[11] << 32) |
+      ((uint64_t)mask_ptr[12] << 24) | ((uint64_t)mask_ptr[13] << 16) |
+      ((uint64_t)mask_ptr[14] << 8) | ((uint64_t)mask_ptr[15]);
+
+  if(mask_high == 0) {
+    // if the high bits are all 0, then low bits must be all 0.
+    return mask_low == 0;
+  }
+
+  // Check that all the 1's are contiguous from left to right
+  uint64_t inv_mask_high = ~mask_high;
+  if ((inv_mask_high + 1) & inv_mask_high) {
+    return 0;
+  }
+
+  // If there was one or more 0's then low has to be all zeros
+  if(inv_mask_high) {
+    return mask_low == 0;
+  }
+
+  // Otherwise low may be all zero's, and if not then the 1's must be contiguous
+  // from left to right
+  return (mask_low == 0 || ((~mask_low + 1) & ~mask_low) == 0);
+}
+
+int validate_cidr_mask(CBS *cidr_mask) {
+  size_t mask_len = CBS_len(cidr_mask);
+  switch (mask_len) {
+    case IPV4_ADDR_LEN:
+      return validate_ipv4_cidr_mask(CBS_data(cidr_mask), mask_len);
+    case IPV6_ADDR_LEN:
+      return validate_ipv6_cidr_mask(CBS_data(cidr_mask), mask_len);
```

Comment:
I was able to cross reference `nc_ip` with OpenSSL's implementation to verify things are correct. Is there a similar implementation here to cross reference or are these two functions new logic?

Reply:
OpenSSL did not handle validation of the mask per this comment: https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/x509v3/v3_ncons.c#L698

But a valid CA shouldn't have name constraints with an invalid CIDR mask/prefix, so I think it makes sense to sanity check that here.